################################################################################
#
#                       Cat Anal Gland Microbiome 
#                      
#         Rojas et al 2023. Characterization of the microbiome and 
#         volatile compounds in anal gland secretions from domestic
#                 cats using metagenomics and metabolomics
#
#                       By: Connie Rojas
#                       Created: 17 July 2022
#                       Last updated: 20 April 2023
#
################################################################################

## CODE FOR:  1) generating stacked barplots showing the abundances of Bacterial 
#                Families and Genera in the Anal Gland

#             2) generating stacked barplots showing the abundances of Bacterial
#               Genera in the Perianal Gland

# Data generated by Kraken2/Bracken, which outputs the taxonomic profiles of 
#            Illumina shotgun sequences using a combination of databases

source(file="scripts/00_configure.R"); #load necessary packages and specifications


################################################################################
#             1.  Load taxa abundance table output by Kraken2/Bracken 
#                           and sample metadata
################################################################################

krakF=read.csv("data/00_bracken_Family.csv", header=T);
load("data/01_bracken_Genus_nocontam.Rdata");
load("data/01_metadata_formatted.Rdata");

# make a vector of your sample names for later
meta2=meta[meta$swab_type=="Anal.Sac",];
samples=meta2$sampleID;


################################################################################
#             2. Create stacked bar plots of Bacterial Family abundances
################################################################################

# sum up taxa abundances 
kfam=krakF[,which(names(krakF) 
                  %in% c(samples, "Family"))];
colnames(kfam)[1]="taxa";
kfam=aggregate(.~taxa,kfam, sum);
kfam=kfam[rowSums(kfam[-1])>1,];

# calculate proportions
kfam[,-1] <- lapply(kfam[,-1], function(x) (x/sum(x))*100);
print(colSums(kfam[-1]));

# keep taxa >1% relative abundance across samples
kfam$AVG=rowMeans(kfam[,-1]);
kfam=kfam[kfam$AVG>1,];
kfam$AVG=NULL;

# denote the rest of Families as "Other"
newrow=c(NA, 100-colSums(kfam[2:ncol(kfam)])); 
kfam=rbind(kfam, newrow); 
kfam$taxa=as.character(kfam$taxa);
kfam[nrow(kfam),1]="Other";

# melt data frame for ggplot
ordbar<-reshape2::melt(kfam, id.vars="taxa",value.name = "abun");
colnames(ordbar)[2]="sampleID";
ordbar=merge(ordbar, meta2, by="sampleID");

# set up color palette
c25 <- c("dodgerblue2", "maroon", "green4", "#6A3D9A", "#FF7F00", 
         "black", "gold1","skyblue2", "#FB9A99", "palegreen2", 
         "#CAB2D6", "#FDBF6F", "gray45", "khaki2","orchid1", 
         "deeppink1", "blue1", "steelblue4", "darkturquoise", 
         "green1", "yellow4", "yellow3","darkorange4", "brown");

# plot
barp1=ggplot(data=ordbar, 
             aes(reorder(sampleID,sample_date),y=abun, fill=taxa))+
  geom_bar(stat = "identity")+
  theme_bw()+ 
  labs(x = "",
       y = "Relative Abundance (%)",
       fill="Bacterial Family",
       title="")+
  scale_fill_manual(values=c25)+
  theme(legend.position="right", 
        legend.text = element_text(size=13),
        legend.title = element_text(size=13, face="bold"),
        legend.key.size = unit(0.9,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_text(size=13, face="bold"),
        axis.text.y = element_text(size=13),
        axis.text.x=element_text(size=10, angle=85, vjust=0.5),
        axis.title.x = element_text(size=13, face="bold")); plot(barp1);

# save image 
ggsave(filename="02_BacterialFamily_composition.pdf",
       device="pdf",path="./figures",
       plot=barp1,
       width=7,
       height=4.5,
       units="in",
       dpi=500);


################################################################################
#             3. Create stacked bar plots of Bacterial Genus abundances           
################################################################################

# sum up taxa abundances 
kgen=krak[,which(names(krak) 
                  %in% c(samples, "Genus"))];
colnames(kgen)[1]="taxa";
kgen=aggregate(.~taxa,kgen, sum);
kgen=kgen[rowSums(kgen[-1])>0,];

# calculate proportions
kgen[,-1] <- lapply(kgen[,-1], function(x) (x/sum(x))*100);
print(colSums(kgen[-1]));
kgen$AVG=rowMeans(kgen[,-1]);

# pick one of these cutoffs (showcase different bacterial groups)
    # a) All taxa >1.75% relative abundance 
kgen2=kgen[kgen$AVG>1.75,]; 

    # b) Taxa with <1.76% relative abundance but > 0.37% relative abundance
kgen2=kgen[kgen$AVG<=1.76,]; 
kgen2=kgen2[kgen2$AVG>0.37,];

    # c) Taxa with <0.4% relative abundance but > 0.18% relative abundance
kgen2=kgen[kgen$AVG<=0.402,]; 
kgen2=kgen2[kgen2$AVG>0.18,];

# delete the average column you created
kgen2$AVG=NULL;

# This step only applies if you selected cutoff a) 
# otherwise SKIP
newrow=c(NA, 100-colSums(kgen2[2:ncol(kgen2)])); 
kgen2=rbind(kgen2, newrow); 
kgen2$taxa=as.character(kgen2$taxa);
kgen2[nrow(kgen2),1]="Other";

# melt data frame for ggplot
genbar<-reshape2::melt(kgen2, id.vars="taxa",value.name = "abun");
colnames(genbar)[2]="sampleID";
genbar=merge(genbar, meta2, by="sampleID");

# set up color palette
fam_col=c("#771155", "#AA4488", "#CC99BB", "#114477","#77AADD", 
           "#117777", "#88CCAA","#774411",
           "#777711",  
           "#DDDD77","#DDAA77","#bdbdbd", "#AA7744", "#AA4455", "grey39", "black");

# plot (do this for each cutoff);
barp2=ggplot(data=genbar, 
             aes(reorder(sampleID,sample_date),y=abun, fill=taxa))+
  geom_bar(stat = "identity")+
  theme_bw()+ 
  labs(x = "",
       y = "Relative Abundance (%)",
       fill="Bacterial Genus",
       title="")+
  scale_fill_manual(values=fam_col)+
  theme(legend.position="right", 
        legend.text = element_text(size=13),
        legend.title = element_text(size=13, face="bold"),
        legend.key.size = unit(0.9,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_text(size=13, face="bold"),
        axis.text.y = element_text(size=13),
        axis.text.x=element_text(size=10, angle=85, vjust=0.5),
        strip.text = element_text(size =11,face="bold"),
        axis.title.x = element_text(size=13, face="bold")); plot(barp2);

# save image 
ggsave(filename="02_BacterialGenus_composition_1.75.pdf",
       device="pdf",path="./figures",
       plot=barp2,
       width=6.4,
       height=4.5,
       units="in",
       dpi=500);


################################################################################
#             4. Create stacked bar plots of Bacterial Genus abundances  in
#                        perianal region vs. anal gland
################################################################################

# select the six cats for which we have perianal data for
meta2=meta[meta$name %in% c("Cannoli","Cheddar","Cooper",
                            "Gabby","Goose","Virginia"),];
samples=meta2$sampleID;

# sum up taxa abundances 
kgen=krak[,which(names(krak) 
                 %in% c(samples, "Genus"))];
colnames(kgen)[1]="taxa";
kgen=aggregate(.~taxa,kgen, sum);
kgen=kgen[rowSums(kgen[-1])>0,];

# calculate proportions
kgen[,-1] <- lapply(kgen[,-1], function(x) (x/sum(x))*100);
print(colSums(kgen[-1]));
kgen$AVG=rowMeans(kgen[,-1]);

# pick one of these cutoffs (showcase different bacterial groups)
# a) All taxa >1.75% relative abundance 
kgen2=kgen[kgen$AVG>1.75,]; 

# b) Taxa with <1.76% relative abundance but > 0.37% relative abundance
kgen2=kgen[kgen$AVG<=1.76,]; 
kgen2=kgen2[kgen2$AVG>0.37,];

# delete the average column you created
kgen2$AVG=NULL;

# This step only applies if you selected cutoff a) 
# otherwise SKIP
newrow=c(NA, 100-colSums(kgen2[2:ncol(kgen2)])); 
kgen2=rbind(kgen2, newrow); 
kgen2$taxa=as.character(kgen2$taxa);
kgen2[nrow(kgen2),1]="Other";

# melt data frame for ggplot
genbar<-reshape2::melt(kgen2, id.vars="taxa",value.name = "abun");
colnames(genbar)[2]="sampleID";
genbar=merge(genbar, meta2, by="sampleID");
genbar$swab_type[genbar$swab_type=="Anal.Sac"]="Anal.Gland";

# plot (do this for each cutoff);
barp3=ggplot(data=genbar, 
             aes(name,y=abun, fill=taxa))+
  geom_bar(stat = "identity")+
  facet_wrap(~swab_type,scales="free")+
  theme_bw()+ 
  labs(x = "",
       y = "Relative Abundance (%)",
       fill="Bacterial Genus",
       title="")+
  scale_fill_manual(values=fam_col)+
  theme(legend.position="right", 
        legend.text = element_text(size=13),
        legend.title = element_text(size=13, face="bold"),
        legend.key.size = unit(0.9,"line"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.y = element_text(size=13, face="bold"),
        axis.text.y = element_text(size=13),
        axis.text.x=element_text(size=10, angle=85, vjust=0.5),
        strip.text = element_text(size =11,face="bold"),
        axis.title.x = element_text(size=13, face="bold")); plot(barp3);

# save image
ggsave(filename="02_Genus_PerianalvAnalGland_1.75.png",
       device="png",path="./figures",
       plot=barp3,
       width=7.4,
       height=5,
       units="in",
       dpi=500);

# calculate means of genera ~ swab type
cdf=data.frame(aggregate(genbar$abun ~ 
                           taxa+swab_type, genbar, mean));
cdf;
